src/services/bookmarks/bookmarks.service.js

```diff
@@ -10,38 +10,53 @@
 angular.module('vlui')
   .service('Bookmarks', function(_, vl, localStorageService, Logger, Dataset) {
     var Bookmarks = function() {
-      this.dict = {};
+      this.list = [];
+      this.annotations = {};
       this.length = 0;
       this.isSupported = localStorageService.isSupported;
     };
 
     var proto = Bookmarks.prototype;
 
     proto.updateLength = function() {
-      this.length = Object.keys(this.dict).length;
+      this.length = this.list.length;
     };
 
     proto.save = function() {
-      localStorageService.set('bookmarks', this.dict);
+      localStorageService.set('bookmarkList', this.list);
     };
 
+    proto.saveAnnotations = function(shorthand) {
+      _.find(this.list, function(bookmark) { return bookmark.shorthand === shorthand; })
+        .chart.annotation = this.annotations[shorthand];
+      this.save();
+    }
+
     proto.load = function() {
-      this.dict = localStorageService.get('bookmarks') || {};
+      this.list = localStorageService.get('bookmarkList') || [];
       this.updateLength();
+
+      // load annotations
+      var annotations = this.annotations;
+      _.forEach(this.list, function(bookmark) {
+        annotations[bookmark.shorthand] = bookmark.chart.annotation;
+      });
     };
 
     proto.clear = function() {
-      this.dict = {};
+      this.list.splice(0, this.list.length);
       this.updateLength();
+      this.annotations = {};
```

Comment:
Why only store the annotation part in the dict?  `isBookmarked` method can also benefit for `O(1)` check with the dictionary.